packages/py-moose-lib/moose_lib/dmv2.py

```diff
@@ -0,0 +1,202 @@
+        raise AttributeError(f"{item} is not a valid column name")
+
+
+class TypedMooseResource(Generic[T]):
+    """Base class for Moose resources that are typed with a Pydantic model."""
+    _t: type[T]
+    name: str
+    columns: Columns[T]
+
+    @classmethod
+    def _get_type(cls, keyword_args: dict):
+        t = keyword_args.get('t')
+        if t is None:
+            raise ValueError(f"Use `{cls.__name__}[T](name='...')` to supply the Pydantic model type`")
+        if not isinstance(t, type) or not issubclass(t, BaseModel):
+            raise ValueError(f"{t} is not a Pydantic model")
+        return t
+
+    def _set_type(self, name: str, t: type[T]):
+        self._t = t
+        self.name = name
+        self.columns = Columns[T](self._t)
+
+    def __class_getitem__(cls, item: type[BaseModel]):
+        def curried_constructor(*args, **kwargs):
+            return cls(t=item, *args, **kwargs)
+
+        return curried_constructor
+
+
+class IngestionFormat(Enum):
+    """Supported formats for data ingestion."""
+    JSON = "JSON"
+    JSON_ARRAY = "JSON_ARRAY"
+
+
+class OlapConfig(BaseModel):
+    """Configuration for OLAP tables."""
+    order_by_fields: list[str] = []
+    deduplicate: bool = True
+
+
+class OlapTable(TypedMooseResource, Generic[T]):
+    config: OlapConfig
+
+    def __init__(self, name: str, config: OlapConfig = OlapConfig(), **kwargs):
+        super().__init__()
+        self._set_type(name, self._get_type(kwargs))
+        self.config = config
+        _tables[name] = self
+
+
+class StreamConfig(BaseModel):
+    """Configuration for data streams."""
+    parallelism: int = 1
+    retention_period: int = 60 * 60 * 24 * 7  # 7 days
+    destination: Optional[OlapTable[Any]] = None
+
+
+@dataclasses.dataclass
+class _RoutedMessage:
+    """Internal class representing a message routed to a specific stream."""
+    destination: "Stream[Any]"
+    values: ZeroOrMany[Any]
+
+
+class Stream(TypedMooseResource, Generic[T]):
+    """
+    A data stream that can transform and route messages.
+    
+    Supports:
+    - Single transformations to other streams
+    - Multi-route transformations where one record can be sent to multiple destinations
+    - Optional connection to a destination table
+    """
+    config: StreamConfig
+    transformations: dict[str, Tuple["Stream[Any]", Callable[[T], ZeroOrMany[Any]]]] = {}
+    _multipleTransformations: Optional[Callable[[T], list[_RoutedMessage]]] = None
+
+    def __init__(self, name: str, config: StreamConfig = StreamConfig(), **kwargs):
+        super().__init__()
+        self._set_type(name, self._get_type(kwargs))
+        self.config = config
+        _streams[name] = self
+
+    def add_transform(self, destination: "Stream[U]", transformation: Callable[[T], ZeroOrMany[U]]):
+        """Add a transformation that sends records to a single destination stream."""
+        self.transformations[destination.name] = (destination, transformation)
```

Comment:
I had to change the ts side like this to use async transformation function. Not sure if we need to do something similar in python

<img width="972" alt="Screenshot 2025-03-20 at 10 49 38 AM" src="https://github.com/user-attachments/assets/ee91a41a-4d57-46d4-b250-356788d556cf" />

Reply:
good catch
i don't think async is as prevalent in python